src/exla/exla_nif_util.h

```diff
@@ -44,6 +58,65 @@ namespace exla {
    */
   ERL_NIF_TERM make(ErlNifEnv* env, std::string &var);
 
+  /*
+   * Templates for resources.
+   */
+  template <typename T>
+  struct resource_object {
+    static ErlNifResourceType *type;
+  };
+  template<typename T> ErlNifResourceType* resource_object<T>::type=0;
+
+  template <typename T>
+  int open_resource(ErlNifEnv* env, const char* mod, const char* name, ErlNifResourceDtor* dtor){
+    ErlNifResourceType *type;
+    ErlNifResourceFlags flags = ErlNifResourceFlags(ERL_NIF_RT_CREATE|ERL_NIF_RT_TAKEOVER);
+    type = enif_open_resource_type(env, mod, name, dtor, flags, NULL);
+    if(type == NULL){
+      resource_object<T>::type = 0;
+      return -1;
+    } else {
+      resource_object<T>::type = type;
+    }
+    return 1;
+  }
+
+  template <typename T>
+  ERL_NIF_TERM get(ErlNifEnv* env, ERL_NIF_TERM term, T* &var){
+    return enif_get_resource(env, term, resource_object<T>::type, (void **) &var);
+  }
+
+  template <typename T>
+  ERL_NIF_TERM make(ErlNifEnv* env, T &var){
+    void* ptr = enif_alloc_resource(resource_object<T>::type, sizeof(T));
+    new(ptr) T(std::move(var));
+    ERL_NIF_TERM ret = enif_make_resource(env, ptr);
```

Comment:
Remember we need to call `enif_release_resource`, so we should probably add a TODO to this function and everywhere else we use enif_make_resource.

Reply:
I'll add it...I added those in and started testing with it earlier and ran into a strange problem. I don't know the underlying implementation of inspecting References, but I assume it shows the memory address of the resource?

Basically, when I added in `enif_release_resource` after `enif_make_resource` I would get valid references returned back, but they weren't valid addresses...sometimes the reference would be: `#Reference<0.0.49482.0>` or something similar that I knew was garbage....